### PR TITLE
add cuda core gemm for m <=4

### PR DESCRIPTION
**Pull request by zhink**
### PR Category

Inference

### PR Types

New features

### Description

pcard-71500
To speed up gemm calculations when m<=4, you should export FLAGS_cuda_core_int8_gemm=1 to use it.
speed data is :
<meta charset="UTF-8"><meta charset="UTF-8"><div class="mp-table-align"><div class="mp-table-wrapper"><div data-slate-node="element" style="padding-left:0px" class="mp-table-container">
m | k | n | L20(秒) | L20(秒) | 加速(%) | V100(秒) | V100(秒) | 加速(%)
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | 11008 | 2048 | 19.46 | 8.98 | 54 | 60.57 | 29.73 | 51
1 | 2048 | 11008 | 16.52 | 9.74 | 41 | 53.18 | 29.33 | 45
1 | 15360 | 5120 | 51.33 | 25.16 | 51 | 178.45 | 93.54 | 48
1 | 5120 | 15360 | 49.49 | 23.28 | 53 | 167.73 | 93.92 | 44
2 | 11008 | 2048 | 20.01 | 9.54 | 52 | 63.71 | 29.67 | 53
2 | 2048 | 11008 | 17.23 | 13.06 | 24 | 53.85 | 30.2 | 44
2 | 15360 | 5120 | 53.30 | 25.51 | 52 | 185.4 | 94.34 | 49
2 | 5120 | 15360 | 51.73 | 26.19 | 49 | 171.33 | 94.88 | 45
3 | 11008 | 2048 | 20.36 | 10.02 | 51 | 64.32 | 30.55 | 53
3 | 2048 | 11008 | 18.17 | 17.59 | 3 | 55.79 | 32.05 | 43
3 | 15360 | 5120 | 54.71 | 26.64 | 51 | 186.69 | 95.12 | 49
3 | 5120 | 15360 | 53.96 | 29.56 | 45 | 173.7 | 96.69 | 44
4 | 11008 | 2048 | 20.87 | 11.13 | 47 | 65.15 | 30.57 | 53
4 | 2048 | 11008 | 19.17 | 22.18 | -16 | 56.78 | 33.54 | 41
4 | 15360 | 5120 | 57.51 | 29.99 | 48 | 188.44 | 96.75 | 49
4 | 5120 | 15360 | 57.63 | 35.84 | 38 | 176.62 | 98.02 | 45

</div><div class="mp-table-serialize-flag"><br/></div></div></div><span data-morpho-doc-data='{"token":"eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwiYXBwSWQiOjEsInVpZCI6IkcwclhMb2E0dXMiLCJkb2NJZCI6Ik0yQ1pTS3FNUFdRQ3ZLIn0..LXkm5x1iJEG3TpDQ.QHC6RZtZVPkzTt4oSpUEYubmnTuIN7Aht5xD2xNEgd9vTKB0srQ3_PipWSTbLPC70wx99nXY3jCpA8jcLGdvf9WEi5SjDcI0fZGcx_HqHVKhOIMH10tuh_AxbcTB7DHsMDdoGWw4331jpHuFtRjjvcXmfbsyMUnxBk8yIxoDODQDIHWjqInpvtz9QJDVo6FABld-rVjjfcDaXfxAgBute5HYQA.uvzEgoETNQx2Z84ZbJK_9g","appId":"1"}' class='mp-morpho-clipboard-doc-data'>﻿</span>

### CODE REVIEW

File: paddle/phi/kernels/fusion/cutlass/cuda_gemm/cuda_gemm.cu [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdio.h>
+#include "cutlass/numeric_conversion.h"
+#include "cuda_gemm.h"
+#include <cub/cub.cuh>
+#define ENABLE_FP8
```
Comment (vivienfanghuagood) [inline paddle/phi/kernels/fusion/cutlass/cuda_gemm/cuda_gemm.cu]: 这个可以默认开启吗？

File: paddle/phi/kernels/fusion/cutlass/cuda_gemm/cuda_gemm.cu [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+#define ENABLE_BF16
```
Comment (vivienfanghuagood) [inline paddle/phi/kernels/fusion/cutlass/cuda_gemm/cuda_gemm.cu]: 这个同上

File: paddle/phi/kernels/impl/matmul_kernel_impl.h [hunk 3]
```diff
@@ -1847,6 +1851,25 @@ MatmulJudgeDtypeKernel(const Context& ctx,
...
+        DenseTensor delta;
+        phi::EmptyKernel<float, Context>(
+            ctx, {y.dims()[1], y.dims()[0]}, DataType::INT8, &delta);
+        phi::TransposeKernel<int8_t, Context>(ctx, y, {1, 0}, &delta);
+        std::vector<int8_t> tmp(delta.numel());
+        cudaMemcpy(tmp.data(),
+                   delta.data<int8_t>(),
+                   sizeof(int8_t) * delta.numel(),
+                   cudaMemcpyDeviceToHost);
+        phi::fusion::cutlass_internal::CudaGemm<T, Context>(ctx, x, delta, out);
```
Comment (vivienfanghuagood) [inline paddle/phi/kernels/impl/matmul_kernel_impl.h]: 为啥这里有deviceToHost的拷贝呢？

**Review by vivienfanghuagood — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for flags

**Pull request merged**